### DISCUSSION TITLE
Native Rust plugins: Capturing configuration in `checkpoint_async`

### DISCUSSION THREAD

**Discussion by invakid404** [apollographql/router#1326] [Q&A] [answered]
Hello everyone!

I'm new to Apollo Router (and to Rust in general), and I'm having a hard time figuring out how to make use of a plugin's configuration within `checkpoint_async`.

I'm testing stuff with the "auth" plugin example. My end goal is to be able to pass the URL of the auth service to the plugin and make use of it within `checkpoint_async`, but I run into various lifetime requirement errors and I have no idea what the proper solution is.

For example, if I try to refer to a configuration property directly within the future returned by `checkpoint_async`, I get the following error:
![image](https://user-images.githubusercontent.com/38114607/176452141-ac655f21-5c4a-435e-bed8-50a61cbce22d.png)

If I attempt to work with a clone of the configuration property, I get this:
![image](https://user-images.githubusercontent.com/38114607/176452377-ff8fb85f-2f50-4018-89e5-7ca2b084c662.png)

By following the hint and forcing the closure to take ownership of `message`, I get yet another lifetime error:
![image](https://user-images.githubusercontent.com/38114607/176452617-c5347acf-c540-4bd7-8a1d-a06a323145da.png)

Attempting to add `move` to the future doesn't help either:
![image](https://user-images.githubusercontent.com/38114607/176452738-7d19771f-008b-48c9-9796-69054ac18dfc.png)

The only thing that I've found to work is if I do yet another clone of `message`, but I really don't feel like this is the proper solution to this problem. While this works, it doesn't sound like a great idea:
```rust
    fn router_service(
        &mut self,
        service: BoxService<
            RouterRequest,
            RouterResponse<BoxStream<'static, ResponseBody>>,
            BoxError,
        >,
    ) -> BoxService<RouterRequest, RouterResponse<BoxStream<'static, ResponseBody>>, BoxError> {
        let message = self.configuration.message.clone();

        ServiceBuilder::new()
            .checkpoint_async(move |request: RouterRequest| {
                let message = message.clone();

                Box::pin(async move {
                    println!("{}", message);

                    // Do some async call here to auth, and decide if to continue or not.
                    Ok(ControlFlow::Continue(request))
                })
            })
            .buffered()
            .service(service)
            .boxed()
    }
```

Does anyone have a clue?

**Comment by bnjjj** [comment, marked as answer]
Actually cloning here is the solution. If it's a big message and you're concerned about cloning you can wrap your message into an `Arc` but it sounds like useless optimization to me.

**Reply by invakid404** [reply]
Thanks for confirming. And no, it's not a big string, so I suppose cloning twice is fine. I'll look into Arcs though, they're a new concept to me.
